rmw_dds_common/src/qos.cpp

```diff
@@ -0,0 +1,243 @@
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_QOS_POLICY_LIVELINESS_UNKNOWN == publisher_qos.liveliness) {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    _write_to_buffer("ERROR: Publisher liveliness is unknown", reason, reason_size);
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+  if (RMW_QOS_POLICY_LIVELINESS_UNKNOWN == subscription_qos.liveliness) {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    _write_to_buffer("ERROR: Subscription liveliness is unknown", reason, reason_size);
+    return RMW_RET_INVALID_ARGUMENT;
+  }
+
+  // Best effort publisher and reliable subscription
+  if (publisher_qos.reliability == RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT &&
+    subscription_qos.reliability == RMW_QOS_POLICY_RELIABILITY_RELIABLE)
+  {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    reason_ss << "ERROR: Best effort publisher and reliable subscription;";
+  }
+
+  // Volatile publisher and transient local subscription
+  if (publisher_qos.durability == RMW_QOS_POLICY_DURABILITY_VOLATILE &&
+    subscription_qos.durability == RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL)
+  {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    reason_ss << "ERROR: Volatile publisher and transient local subscription;";
+  }
+
+  const rmw_time_t & pub_deadline = publisher_qos.deadline;
+  const rmw_time_t & sub_deadline = subscription_qos.deadline;
+  const rmw_time_t deadline_default = RMW_QOS_DEADLINE_DEFAULT;
+
+  // No deadline for publisher and deadline for subscription
+  if (pub_deadline == deadline_default && sub_deadline != deadline_default) {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    reason_ss << "ERROR: Subscription has a deadline, but publisher does not;";
+  }
+
+  // Subscription deadline is less than publisher deadline
+  if (pub_deadline != deadline_default && sub_deadline != deadline_default) {
+    if (sub_deadline < pub_deadline) {
+      *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+      reason_ss << "ERROR: Subscription deadline is less than publisher deadline " <<
+        "(" << sub_deadline << " < " << pub_deadline << ");";
+    }
+  }
+
+  // Automatic liveliness for publisher and manual by topic for subscription
+  if (publisher_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_AUTOMATIC &&
+    subscription_qos.liveliness == RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC)
+  {
+    *compatibility = RMW_QOS_COMPATIBILITY_ERROR;
+    reason_ss << "ERROR: Publisher's liveliness is automatic and subscription's liveliness is " <<
+      "manual by topic;";
+  }
+
+  const rmw_time_t & pub_lease = publisher_qos.liveliness_lease_duration;
+  const rmw_time_t & sub_lease = subscription_qos.liveliness_lease_duration;
+  const rmw_time_t lease_default = RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT;
+
+  // No lease duration for publisher and lease duration for subscription
+  if (pub_lease == lease_default && sub_lease != lease_default) {
```

Comment:
I'm not so sure of this.
IIUC, the default is the [participant liveliness lease duration](https://community.rti.com/static/documentation/connext-dds/6.0.1/doc/api/connext_dds/api_cpp/structDDS__DiscoveryConfigQosPolicy.html#a78c4d3b8bda361a00e123f58ea496e4c), so the profiles could still be compatible (with a suitable participant liveliness lease duration).

Reply:
The RMW docs claim that this default value implies liveliness is not tracked at all: https://github.com/ros2/rmw/blob/ae0553adb574c031343965195ec4a73039e62389/rmw/include/rmw/types.h#L433-L434

I haven't confirmed if that is true though.

Reply:
For rmw_cyclonedds, the value of zero (the default) results in an ["infinite" duration](https://github.com/ros2/rmw_cyclonedds/blob/d024823043504ea40af24bf22365a21cd203df55/rmw_cyclonedds_cpp/src/rmw_node.cpp#L1709-L1712).

For rmw_fastrtps and rmw_connext, we seem to be defaulting to their defaults, only setting a value if it is not the rmw default:

https://github.com/ros2/rmw_fastrtps/blob/901b0ed7d286b34cda72b3a1138c6e05bf1096c4/rmw_fastrtps_shared_cpp/src/qos.cpp#L125

https://github.com/ros2/rmw_connext/blob/fd3e6a71fee0f7eb26f84505d16169ce489d6235/rmw_connext_shared_cpp/src/qos.cpp#L119

Reply:
I took a look at the values default values for the talker demo with `ros2 topic info --verbose /chatter`, and the RMWs report very large default values (which I think effectively means liveliness is not tracked):

Value for Fast-DDS and CycloneDDS: 9223372036854775807 nanoseconds
Value for Connext: 2147483649147483647 nanoseconds

Reply:
Liveliness is always somehow tracked, but it seems that the matching rules are fine.